psqlgraph/node.py

```diff
@@ -204,6 +215,58 @@ def edge_predicate(e):
                     queue.append((n, depth + 1))
                     marked.add(n.node_id)
 
+    def _dfs(self, edge_predicate=None, max_depth=None, edge_pointer="in"):
+        """
+        Perform a DFS, with `self` being the root node
+
+        To implement max depth, some node are visited more than once to update shortest
+        path. But those node should only be yield once.
+
+        :param edge_predicate: a predicate performed on an `edge` object in
+            order to decided whether to walk that edge or not
+        :type edge_predicate: func
+        :param max_depth: maximum distance to traverse
+        :type max_depth: int
+        :param edge_pointer: possible values `in`, `out`
+                            `in`: use node.edges_in, default behavior
+                            `out`: use edges_out
+        :type edge_pointer: str
+
+        :return: generator
+        """
+        StackItem = namedtuple('StackItem', ['node', 'next_child', 'level'])
+        edge_predicate = edge_predicate if callable(edge_predicate) else lambda _: True
+        max_depth = float('inf') if max_depth is None else max_depth
+
+        visited = {self.node_id: 0}
+        yield self
+        stack = [StackItem(self, 0, 0)]
+
+        while stack:
+            node, next_child, level = stack.pop()
+            if level >= max_depth:
+                continue
+            edges = node.edges_out if edge_pointer == "out" else node.edges_in
+
+            for j in range(next_child, len(edges)):
+                edge = edges[j]
+                n = edge.dst if edge_pointer == "out" else edge.src
+                if not edge_predicate(edge):
+                    continue
+
+                if n.node_id in visited:
+                    if max_depth == float('inf') or level + 1 >= visited[n.node_id]:
+                        continue
+                    # update levels for max_depth if shorter path found
+                    # but do not yield node again
+                else:
+                    yield n
```

Comment:
Ah! Cuz I hate nesting lol. Just a suggestion tho.

```suggestion
                if n.node_id not in visited:
                    yield n
                    
                elif max_depth == float('inf') or level + 1 >= visited[n.node_id]:
                    continue
```

Reply:
Less nesting probably is always better.